crates/steel-parser/src/lexer.rs

```diff
@@ -802,6 +869,88 @@ mod lexer_tests {
         );
     }
 
+    #[test]
+    fn test_unicode_escapes() {
+        let mut s = TokenStream::new(
+            r#"  #\xAb #\u{0D300} #\u0540 "\x00D;" "\u1044;" "\u{045}"  "#,
+            true,
+            SourceId::none(),
+        );
+
+        assert_eq!(
+            s.next().unwrap(),
+            Token {
+                ty: CharacterLiteral('«'),
+                source: r#"#\xAb"#,
+                span: Span::new(2, 7, SourceId::none())
+            }
+        );
+
+        assert_eq!(
+            s.next().unwrap(),
+            Token {
+                ty: CharacterLiteral('팀'),
+                source: r#"#\u{0D300}"#,
+                span: Span::new(8, 18, SourceId::none())
+            }
+        );
+
+        assert_eq!(
+            s.next().unwrap(),
+            Token {
+                ty: CharacterLiteral('Հ'),
+                source: r#"#\u0540"#,
+                span: Span::new(19, 26, SourceId::none())
+            }
+        );
+
+        assert_eq!(
+            s.next().unwrap(),
+            Token {
+                ty: StringLiteral(Arc::from("\r".to_string())),
+                source: r#""\x00D;""#,
+                span: Span::new(27, 35, SourceId::none())
+            }
+        );
+
+        assert_eq!(
+            s.next().unwrap(),
+            Token {
+                ty: StringLiteral(Arc::from("၄".to_string())),
+                source: r#""\u1044;""#,
+                span: Span::new(36, 45, SourceId::none())
+            }
+        );
+
+        assert_eq!(
+            s.next().unwrap(),
+            Token {
+                ty: StringLiteral(Arc::from("E".to_string())),
+                source: r#""\u{045}""#,
+                span: Span::new(46, 55, SourceId::none())
+            }
+        );
+    }
+
+    #[test]
+    fn test_invalid_unicode_escapes() {
+        let tokens = [
+            r#" #\xd820 "#,
+            r#" #\u{1 "#,
+            r#" "\xabx" "#,
+            r#" "\u0045" "#,
+            r#" #\xaaaaaaaa " "#,
+            r#" "\u{ffffffff}" "#,
+            r#" #\u{} "#,
+        ];
+
+        for token in tokens {
+            let mut s = TokenStream::new(token, true, SourceId::none());
+
+            assert_eq!(s.next().unwrap().ty, Error, "{:?} should be invalid", token);
```

Comment:
I'd like to revisit tokenization errors at some point. They surface very little info in error messages.